UM/CentralFileStorage.py

```diff
@@ -5,101 +5,126 @@
-            os.remove(file_path)
-            Logger.info(f"{file_id} {str(version)} was already stored centrally. Removing duplicate.")
+            if os.path.getsize(path) != os.path.getsize(storage_path):  # As quick check if files are the same, check their file sizes.
+                raise FileExistsError(f"Central file storage already has an item (file or directory) with ID {path_id} and version {str(version)}, but it's different.")
+            new_item_hash = cls._hashItem(path)
+            stored_item_hash = cls._hashItem(storage_path)
+            if new_item_hash != stored_item_hash:
+                raise FileExistsError(f"Central file storage already has an item (file or directory) with ID {path_id} and version {str(version)}, but it's different.")
+            if os.path.isfile(path):
+                os.remove(path)
+            elif os.path.isdir(path):
+                shutil.rmtree(path)
+            Logger.info(f"{path_id} {str(version)} was already stored centrally. Removing duplicate.")
         else:
-            shutil.move(file_path, storage_path)
-            Logger.info(f"Storing new file {file_id} {str(version)}.")
+            shutil.move(path, storage_path)
+            Logger.info(f"Storing new item {path_id}.{str(version)}.")
 
     @classmethod
-    def retrieve(cls, file_id: str, sha256_hash: str, version: Version = Version("1.0.0")) -> str:
+    def retrieve(cls, path_id: str, sha256_hash: str, version: Version = Version("1.0.0")) -> str:
         """
-        Retrieve the file path of a previously stored file.
-        :param file_id: The name of the file to retrieve.
-        :param sha256_hash: A SHA-256 hash of the file you expect to receive from the central storage.
-        :param version: The version number of the file to retrieve.
-        :return: A path to the location of the centrally stored file.
-        :raises FileNotFoundError: There is no file stored with that name and version.
-        :raises IOError: The hash of the file is incorrect. Opening this file could be a security risk.
+        Retrieve the path of a previously stored item (file or directory).
+        :param path_id: The name of the item (file or directory) to retrieve.
+        :param sha256_hash: A SHA-256 hash of the item (file or directory) you expect to receive from the central storage.
+        :param version: The version number of the item (file or directory) to retrieve.
+        :return: A path to the location of the centrally stored item (file or directory).
+        :raises FileNotFoundError: There is no item (file or directory) stored with that name and version.
+        :raises IOError: The hash of the item (file or directory) is incorrect. Opening this item could be a security
+        risk.
         """
-        storage_path = cls._getFilePath(file_id, version)
+        if path_id + str(version) in cls._unmoved_files:
+            return cls._unmoved_files[path_id + str(version)]
```

Comment:
This is a security risk, because it is returning a path here without checking the hash. An attacker could replace the file in this position with a malicious file, and it would be trusted.

I suggest you modify the `_getItemPath` function to include this case, and then execute the rest of the function as normal.

Reply:
That's not true. If a file is not moved it is done because it's bundled (and thus in the same location as where Cura is installed). We skip all checks for bundled plugins, so we should also be able to skip this as far as I can see.

Reply:
As discussed, this is indeed not an attack vector if the `store` function is used correctly.